### ISSUE TITLE
CS8618 warning analysis ignores exceptional path in some cases

### ISSUE THREAD

**Issue by am11** [dotnet/roslyn#44046]
**Version Used**: 

```
csc: 3.2.0-beta2-19303-01 (c9689b7a)
dotnet: 3.1.100
```

**Steps to Reproduce**:

Build the following program with `dotnet build /p:Nullable=enable`

```c#
class X
{
   static bool Condition() => true;

   string fieldA;

   X() // CS8618 reported on this line
   {
       if (Condition()) fieldA = string.Empty;
       if (fieldA == default) throw new System.Exception("invalid setup");
   }
}
```

**Expected Behavior**:

No warning.

**Actual Behavior**:

> X.cs(7,4): warning CS8618: Non-nullable field 'fieldA' is uninitialized. Consider declaring the field as nullable.

Replacing `if (fieldA == default)` with `else` fixes the warning.

**Comment by jcouv** [comment]
Confirmed. Tagging @cston

**Comment by jcouv** [comment]
Here's another scenario (https://github.com/dotnet/roslyn/issues/43523#issuecomment-626255249 @stephentoub) which may motivate merging uninitialized analysis of non-nullable fields and nullable analysis.

----
Here's another form of this that's problematic.  Our real case of this is a bit more complex, but essentially it boils down to this:
```C#
#nullable enable
using System;
public class C {
    private object _value;
    
    public C()
    {
        if (_value is null)
            throw new Exception();
    }
}
```
The compiler should be able to determine that there's no way to exit C's ctor with _value being null, yet it still warns that _value may be uninitialized.

**Comment by RikkiGibson** [comment]
This will be fixed by #46441
